Cabal/src/Distribution/Simple/GHC/Build/ExtraSources.hs

```diff
@@ -191,52 +201,53 @@ buildExtraSources description componentSourceGhcOptions wantDyn viewSources ghcP
+                -- We suffix shared objects with .dyn_o and profiled ones with .p_o.
+                CLib _lib
+                  -- Unless for repl, in which case we only need the vanilla way
+                  | BuildRepl _ <- buildingWhat ->
+                      [vanillaSrcOpts]
+                  | otherwise ->
+                      [vanillaSrcOpts]
+                      ++ [sharedSrcOpts{ghcOptObjSuffix = toFlag "dyn_o"} | wantDyn && (forceSharedLib || withSharedLib lbi)]
+                      ++ [profSrcOpts{ghcOptObjSuffix = toFlag "p_o"} | withProfLib lbi]
 
-          -- For foreign libraries, we determine with which options to build the
-          -- objects (vanilla vs shared vs profiled)
-          CFLib flib
-            | withProfExe lbi -> -- It doesn't sound right to query "ProfExe" for a foreign library...
-                compileIfNeeded profSrcOpts
-            | withDynFLib flib && wantDyn ->
-                compileIfNeeded sharedSrcOpts
-            | otherwise ->
-                compileIfNeeded vanillaSrcOpts
-          -- For the remaining component types (Exec, Test, Bench), we also
-          -- determine with which options to build the objects (vanilla vs shared vs
-          -- profiled), but predicate is the same for the three kinds.
-          _exeLike
-            | withProfExe lbi ->
-                compileIfNeeded profSrcOpts
-            | withDynExe lbi && wantDyn ->
-                compileIfNeeded sharedSrcOpts
-            | otherwise ->
-                compileIfNeeded vanillaSrcOpts
+                -- For foreign libraries, we determine with which options to build the
+                -- objects (vanilla vs shared vs profiled)
+                CFLib flib
+                  | withProfExe lbi -> -- It doesn't sound right to query "ProfExe" for a foreign library...
+                      [profSrcOpts]
+                  | withDynFLib flib && wantDyn ->
+                      [sharedSrcOpts]
+                  | otherwise ->
+                      [vanillaSrcOpts]
+                -- For the remaining component types (Exec, Test, Bench), we also
+                -- determine with which options to build the objects (vanilla vs shared vs
+                -- profiled), but predicate is the same for the three kinds.
+                _exeLike
+                  | withProfExe lbi ->
+                      [profSrcOpts]
+                  | withDynExe lbi && wantDyn ->
+                      [sharedSrcOpts]
+                  | otherwise ->
+                      [vanillaSrcOpts]
+        forM_ allOpts $ \opts -> do
+          files <- fmap concat $ forM sourceFiles $ \sourceFile -> do
+            needsRecomp <- checkNeedsRecompilation sourceFile opts
+            return [sourceFile | needsRecomp]
+          when (not (null files)) $
+            runGhcProg opts { ghcOptInputFiles = toNubListR files }
```

Comment:
Right, that is much better than doing one ghc invocation for each extra source. That alone should already be performance significant when compiling multiple C sources, let alone eventually building all those in parallel.